src/katsdpmodels/models.py

```diff
@@ -0,0 +1,263 @@
+################################################################################
+# Copyright (c) 2020, National Research Foundation (SARAO)
+#
+# Licensed under the BSD 3-Clause License (the "License"); you may not use
+# this file except in compliance with the License. You may obtain a copy
+# of the License at
+#
+#   https://opensource.org/licenses/BSD-3-Clause
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+"""Base functionality common to all model types."""
+
+from abc import ABC, abstractmethod
+from datetime import datetime, timezone
+import io
+import urllib.parse
+import h5py
+from typing import Optional, Union, Any, ClassVar, Type, TypeVar, overload
+
+import numpy as np
+import strict_rfc3339
+
+
+_E = TypeVar('_E', bound='ModelError')
+_M = TypeVar('_M', bound='Model')
+_H = TypeVar('_H', bound='SimpleHDF5Model')
+
+
+class ModelError(ValueError):
+    """A model was found, but the content was incorrect."""
+
+    def __init__(self, *args) -> None:
+        super().__init__(*args)
+        self.url: Optional[str] = None
+        self.original_url: Optional[str] = None
+
+    @classmethod
+    def with_urls(cls: Type[_E], *args,
+                  url: Optional[str] = None, original_url: Optional[str] = None) -> _E:
+        exc = cls(args)
+        exc.url = url
+        exc.original_url = original_url
+        return exc
+
+
+class FileTypeError(ModelError):
+    """The file type (as determined by extension) was not recognised."""
+
+
+class ModelTypeError(ModelError):
+    """The ``model_type`` attribute was missing or did not match the expected value."""
+
+
+class ModelFormatError(ModelError):
+    """The ``model_format`` attribute was missing or did match a known value."""
+
+
+class DataError(ModelError):
+    """The model was missing some data or it had the wrong format."""
+
+
+class ChecksumError(DataError):
+    """The model did not match the checksum embedded in the filename."""
+
+
+class TooManyAliasesError(ModelError):
```

Comment:
Is there a broken alias scenario ?

Reply:
In that case you get a requests HTTP exception, just the same as if you'd specified a non-existent URL yourself.